### PR TITLE
fix(gnovm): improve error message for nil assignment in variable declaration

### PR DESCRIPTION
**Pull request by omarsy**
…aration

<details><summary>Contributors' checklist...</summary>

- [ ] Added new tests, or not needed, or not feasible
- [ ] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [ ] Updated the official documentation or not needed
- [ ] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [ ] Added references to related issues and PRs
- [ ] Provided any useful hints for running manual tests
</details>

### CODE REVIEW

File: gnovm/tests/files/assign29.gno [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package main
+
+func main() {
+	a := 1
+	a = nil
+	println(a)
+}
+
+// Error:
+// main/files/assign29.gno:5:2: cannot use nil as int value in variable declaration
```
Comment (omarsy) [inline gnovm/tests/files/assign29.gno:10]: should be `cannot use nil as int value in assignment`

File: gnovm/pkg/gnolang/type_check.go [hunk 2]
```diff
@@ -289,7 +289,7 @@ func checkAssignableTo(xt, dt Type, autoNative bool) error {
 	// case0
 	if xt == nil { // see test/files/types/eql_0f18
 		if !maybeNil(dt) {
-			panic(fmt.Sprintf("invalid operation, nil can not be compared to %v", dt))
+			panic(fmt.Sprintf("cannot use nil as %v value in variable declaration", dt))
```
Comment (ltzmaxwell) [inline gnovm/pkg/gnolang/type_check.go:292]: thank you for tackling this.

**Comment by ltzmaxwell** [comment]
can you add some more tests like `a + nil`, etc. ?  I guess there would be no more side effect but it's still worth to be more clear.

File: gnovm/pkg/gnolang/type_check_test.go [hunk 3]
```diff
@@ -7,6 +7,7 @@ func TestCheckAssignableTo(t *testing.T) {
 
 	tests := []struct {
 		name       string
+		n          Node
```
Comment (thehowl) [inline gnovm/pkg/gnolang/type_check_test.go:10]: if this is always nil, just pass `nil` below

File: gnovm/pkg/gnolang/types.go [hunk 4]
```diff
@@ -2427,7 +2427,7 @@ func IsImplementedBy(it Type, ot Type) bool {
 // specTypeval is Type if spec is TypeKind.
 // NOTE: type-checking isn't strictly necessary here, as the resulting lookup
 // map gets applied to produce the ultimate param and result types.
-func specifyType(store Store, lookup map[Name]Type, tmpl Type, spec Type, specTypeval Type) {
+func specifyType(n Node, store Store, lookup map[Name]Type, tmpl Type, spec Type, specTypeval Type) {
```
Comment (thehowl) [inline gnovm/pkg/gnolang/types.go:2430]: maybe put Node after Store

**Review by thehowl — APPROVED** [review]
leaving @ltzmaxwell for final review/merge

File: gnovm/pkg/gnolang/type_check_test.go (hunk 3 above)
Comment (omarsy) [reply gnovm/pkg/gnolang/type_check_test.go:10]: Yes right my bad

File: gnovm/pkg/gnolang/types.go (hunk 4 above)
Comment (omarsy) [reply gnovm/pkg/gnolang/types.go:2430]: right

Comment (omarsy) [reply gnovm/pkg/gnolang/types.go]: done here https://github.com/gnolang/gno/pull/3068/commits/7ae5d59fd3ba2303cc25c4e807a8ae6f685c617c

File: gnovm/pkg/gnolang/type_check_test.go (hunk 3 above)
Comment (omarsy) [reply gnovm/pkg/gnolang/type_check_test.go]: done here https://github.com/gnolang/gno/pull/3068/commits/7ae5d59fd3ba2303cc25c4e807a8ae6f685c617c

File: gnovm/pkg/gnolang/preprocess.go [hunk 5]
```diff
@@ -2399,9 +2400,9 @@ func defineOrDecl(
 	tvs := make([]TypedValue, numNames)
 
 	if numVals == 1 && numNames > 1 {
-		parseMultipleAssignFromOneExpr(sts, tvs, store, bn, nameExprs, typeExpr, valueExprs[0])
+		parseMultipleAssignFromOneExpr(sts, tvs, store, bn, n, nameExprs, typeExpr, valueExprs[0])
```
Comment (ltzmaxwell) [inline gnovm/pkg/gnolang/preprocess.go:2403]:
```suggestion
		parseMultipleAssignFromOneExpr(store, bn, n, nameExprs, typeExpr, valueExprs[0], sts, tvs)
```

File: gnovm/pkg/gnolang/preprocess.go [hunk 6]
```diff
@@ -2399,9 +2400,9 @@ func defineOrDecl(
...
 	} else {
-		parseAssignFromExprList(sts, tvs, store, bn, isConst, nameExprs, typeExpr, valueExprs)
+		parseAssignFromExprList(sts, tvs, store, bn, n, isConst, nameExprs, typeExpr, valueExprs)
```
Comment (ltzmaxwell) [inline gnovm/pkg/gnolang/preprocess.go:2405]:
```suggestion
		parseAssignFromExprList(store, bn, n, isConst, nameExprs, typeExpr, valueExprs, sts, tvs)
```

Comment (ltzmaxwell) [reply gnovm/pkg/gnolang/preprocess.go:2405]: please also check the func impl.

File: gnovm/pkg/gnolang/preprocess.go [hunk 7]
```diff
@@ -3491,14 +3494,14 @@ func isConstType(x Expr) bool {
 }
 
 // check before convert type
-func checkOrConvertType(store Store, last BlockNode, x *Expr, t Type, autoNative bool) {
+func checkOrConvertType(store Store, last BlockNode, n Node, x *Expr, t Type, autoNative bool) {
 	if debug {
 		debug.Printf("checkOrConvertType, *x: %v:, t:%v \n", *x, t)
```
Comment (ltzmaxwell) [inline gnovm/pkg/gnolang/preprocess.go:3499]: will be good also modify this.

File: gnovm/pkg/gnolang/type_check.go [hunk 8]
```diff
@@ -282,7 +282,7 @@ func checkValDefineMismatch(n Node) {
 // Assert that xt can be assigned as dt (dest type).
 // If autoNative is true, a broad range of xt can match against
 // a target native dt type, if and only if dt is a native type.
-func checkAssignableTo(xt, dt Type, autoNative bool) error {
+func checkAssignableTo(n Node, xt, dt Type, autoNative bool) error {
 	if debug {
 		debug.Printf("checkAssignableTo, xt: %v dt: %v \n", xt, dt)
```
Comment (ltzmaxwell) [inline gnovm/pkg/gnolang/type_check.go:287]: same too.

**Review by ltzmaxwell — CHANGES_REQUESTED** [review]
LGTM, thank you for fixing this.

**Review by ltzmaxwell — APPROVED** [review]
